VMHyperV: Added the ability to enable or disable the TPM on a VM

jmerriweather opened the pull request:
#### Pull Request (PR) description
This adds the ability to enable or disable the TPM on a virtual machine

#### This Pull Request (PR) fixes the following issues

- Fixes #214 

#### Task list

- [X] Added an entry to the change log under the Unreleased section of the file CHANGELOG.md.
      Entry should say what was changed and how that affects users (if applicable), and
      reference the issue being resolved (if applicable).
- [ ] Resource documentation updated in the resource's README.md.
- [X] Resource parameter descriptions updated in schema.mof.
- [X] Comment-based help updated, including parameter descriptions.
- [ ] Localization strings added/updated in all localization files as appropriate.
- [X] Examples appropriately added/updated.
- [X] Unit tests added/updated. See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [ ] Integration tests added/updated (where possible). See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [X] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dsccommunity/HyperVDsc/215)
<!-- Reviewable:end -->

jmerriweather committed:
Added the ability to enable or disable the TPM on a VM

jmerriweather committed:
Added the ability to enable or disable the TPM on a VM

jmerriweather committed:
Fixed module defaults, updated module schema.mof, added example

jmerriweather committed:
Fixed module defaults, updated module schema.mof, added example

jmerriweather committed:
Initialise the VMKeyProtector if it is not initialised

jmerriweather committed:
Initialise the VMKeyProtector if it is not initialised

johlju committed:
Merge branch 'main' into enable-vm-tpm

johlju commented:
@jmerriweather look through coderabbitai comments, the first two I read was relevant to resolve, please look at the others as well.

jmerriweather committed:
Resolve PR #215 review feedback for TPM support

- Fix inverted Enable/Disable-VMTPM logic in Set-TargetResource existing-VM path
- Return EnableTPM (was TpmEnabled) from Get-TargetResource to match schema
- Fix broken TPM unit tests (correct -EnableTPM param, mock Test-VMTpmEnabled,
  correct expectations) and add missing Get-VMSecurity/Test-VMTpmEnabled mocks
  so existing Gen2 tests no longer hit the unimplemented stub
- Link CHANGELOG entry to issue #214 and expand example comment-based help

jmerriweather committed:
Fix TPM key protector brace style

jmerriweather committed:
Add inputs and outputs to TPM example help

jmerriweather committed:
Enable TPM without prompting for existing generation 2 VMs in Set-TargetResource

jmerriweather committed:
Enhance Set-TargetResource to handle TPM configuration for Generation 2 VMs

johlju reviewed (changes requested):
> @johlju reviewed 5 files and all commit messages, and made 4 comments.
> *[Reviewable](https://reviewable.io/reviews/dsccommunity/hypervdsc/215)* status: all files reviewed, 10 unresolved discussions (waiting on jmerriweather).

___
*[`source/DSCResources/DSC_VMHyperV/DSC_VMHyperV.psm1` line 441 at r1](https://reviewable.io/reviews/dsccommunity/hypervdsc/215#-OzQkvA1EXPuUDTYKA7O:-OzQkvA1EXPuUDTYKA7P:bam0n4e) ([raw file](https://github.com/dsccommunity/hypervdsc/blob/8a2f282b4e23c2e2fa5f4df9bccce644b87048cf/source/DSCResources/DSC_VMHyperV/DSC_VMHyperV.psm1#L441)):*
> ```PowerShell
> 
>                 # Retrive the current TPM state
>                 $vmTPMEnabled = Test-VMTpmEnabled -Name $Name
> ```

We should probably call this only if `$vmobj.Generation -eq 2`? Also only call it if we actually should enforce TPM, see other comment.



_Code quote:_
```PowerShell
$vmTPMEnabled = Test-VMTpmEnabled -Name $Name
```
___
*[`source/DSCResources/DSC_VMHyperV/DSC_VMHyperV.psm1` line 442 at r1](https://reviewable.io/reviews/dsccommunity/hypervdsc/215#-OzQkWKJ3uaMx593vClF:-OzQkWKK449U0tuA_KGT:b-lgetpt) ([raw file](https://github.com/dsccommunity/hypervdsc/blob/8a2f282b4e23c2e2fa5f4df9bccce644b87048cf/source/DSCResources/DSC_VMHyperV/DSC_VMHyperV.psm1#L442)):*
> ```PowerShell
>                 # Retrive the current TPM state
>                 $vmTPMEnabled = Test-VMTpmEnabled -Name $Name
>                 if ($EnableTPM -ne $vmTPMEnabled)
> ```

We should only enforce it if $PSBoundParameters contains the key (user passed the parameter), otherwise we could wrongly revert a exiting VM to not have TPM enabled if the user does not pass the parameter.



_Code quote:_
```PowerShell
if ($EnableTPM -ne $vmTPMEnabled)
```
___
*[`source/DSCResources/DSC_VMHyperV/DSC_VMHyperV.psm1` line 942 at r1](https://reviewable.io/reviews/dsccommunity/hypervdsc/215#-OzQlpa95UGdmAc1blLB:-OzQlpa95UGdmAc1blLC:b-4msvrs) ([raw file](https://github.com/dsccommunity/hypervdsc/blob/8a2f282b4e23c2e2fa5f4df9bccce644b87048cf/source/DSCResources/DSC_VMHyperV/DSC_VMHyperV.psm1#L942)):*
> ```PowerShell
> 
>                 $vmTPMEnabled = Test-VMTpmEnabled -Name $Name
>                 if ($EnableTPM -ne $vmTPMEnabled)
> ```

Same as previous comments, only enforce if user pass parameter.
___
*[`source/DSCResources/DSC_VMHyperV/DSC_VMHyperV.psm1` line 57 at r4](https://reviewable.io/reviews/dsccommunity/hypervdsc/215#-OzQmVdM89ammzhkh_ou:-OzQmVdM89ammzhkh_ov:b-5xr9g) ([raw file](https://github.com/dsccommunity/hypervdsc/blob/51d847d507a1d9b65900f3ce70521eb9c57253c4/source/DSCResources/DSC_VMHyperV/DSC_VMHyperV.psm1#L57)):*
> ```PowerShell
> 
>         # Retrieve TPM status (can only be enabled on Generation 2 VMs) and return boolean.
>         $vmTPMState = ($vmobj | Get-VMSecurity).TpmEnabled
> ```

We should probably use `Test-VMTpmEnabled -Name $Name` here too?



_Code quote:_
```PowerShell
($vmobj | Get-VMSecurity).TpmEnabled
```


<!-- Sent from Reviewable.io -->
